### PR TITLE
Standardized protobuf update code style

### LINKED ISSUE #1557: Set protobuf fields by copying to `mutable_...` and `add_...` instead of using `set_allocated_...`
### Description of the task
as titled
The goal is to avoid using release in some places and not in others, since that can be confusing to new (and old) members.

Old, error-prone style:
```
std::unique_ptr<sub_proto> sub_p_ptr;
std::unique_ptr<proto_to_add> add_p_ptr;
// setting a field in proto
proto.set_allocated_sub_proto(sub_p_ptr.release());
// Adding add_p_ptr to a repeated field
*(proto.add_sub_proto()) = *add_p_ptr;
```
New convention:
```
std::unique_ptr<sub_proto> sub_p_ptr;
std::unique_ptr<proto_to_add> add_p_ptr;
// setting a field in proto
*(proto.mutable_sub_proto()) = *sub_p_ptr;
// Adding add_p_ptr to a repeated field
*(proto.add_sub_proto()) = *add_p_ptr;
```

### Acceptance criteria

- [ ] change instances
- [ ] change code conventions

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
The goal is to avoid using release in some places and not in others, since that can be confusing to new (and old) members.

A key motivation is that we can't use release() with repeated fields, but arguably we should avoid release if possible since we try to avoid raw pointers in general

Old, error-prone style:
```
std::unique_ptr<sub_proto> sub_p_ptr;
std::unique_ptr<proto_to_add> add_p_ptr;
// setting a field in proto
proto.set_allocated_sub_proto(sub_p_ptr.release());
// Adding add_p_ptr to a repeated field
*(proto.add_sub_proto()) = *add_p_ptr;
```
New convention:
```
std::unique_ptr<sub_proto> sub_p_ptr;
std::unique_ptr<proto_to_add> add_p_ptr;
// setting a field in proto
*(proto.mutable_sub_proto()) = *sub_p_ptr;
// Adding add_p_ptr to a repeated field
*(proto.add_sub_proto()) = *add_p_ptr;
```

### Testing Done
no CI regressions

### Resolved Issues
resolve https://github.com/UBC-Thunderbots/Software/issues/1557

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: docs/code-style-guide.md [hunk 1]
```diff
@@ -390,3 +390,22 @@ Some general guidelines when writing tests are:
 ### Protobuf
 
 Protobufs that we define should follow [Google's Protobuf Style Guide](https://developers.google.com/protocol-buffers/docs/style).
+
+* When initializing a protobuf from a unique pointer, avoid passing the `release()`'d pointer to set\_allocated and AddAllocated. Releasing unique pointers makes it too easy to cause memory leaks.
```
Comment (akhilveeraghanta) [inline docs/code-style-guide.md]:
```suggestion
* When initializing a protobuf from a unique pointer, avoid passing the `release()`'d pointer to `set_allocated` and `AddAllocated`. Releasing unique pointers makes it too easy to cause memory leaks.
```

File: docs/code-style-guide.md [hunk 2]
```diff
@@ -390,3 +390,22 @@ Some general guidelines when writing tests are:
...
+
+  // Incorrect
+  vision_msg->set_allocated_ball_state(createBallStateMsg(world.ball()).release());
+  segments.AddAllocated(segment_1.release());
+  *(sensor_msg_1.add_tbots_robot_msgs())   = *tbots_robot_msg_id_1;
+
+  // Correct
+  *(vision_msg->mutable_ball_state()) = *createBallStateMsg(world.ball());
+  *(segments.Add()) = *segment_1;
+  *(sensor_msg_1.add_tbots_robot_msgs())   = *tbots_robot_msg_id_1;
```
Comment (akhilveeraghanta) [inline docs/code-style-guide.md]: would be useful to indicate which is which

i.e
Nested Message
not sure what the middle one is
Repeated Message

Comment (jonathanlew) [reply docs/code-style-guide.md]: there's context at the top of the example block

Comment (akhilveeraghanta) [reply docs/code-style-guide.md]: oh lol totally missed them

**Review by akhilveeraghanta — DISMISSED** [review]
✔️

**Review by kelvinkoon — DISMISSED** [review]
Really like this, good work! :tada:

File: docs/code-style-guide.md [hunk 3]
```diff
@@ -390,3 +390,22 @@ Some general guidelines when writing tests are:
...
+* When initializing a protobuf from a unique pointer, avoid passing the `release()`'d pointer to `set_allocated` and `AddAllocated`. Releasing unique pointers makes it too easy to cause memory leaks.
+  ```cpp
+  /** Context:
+   * ball_state is field of vision_msg
+   * segments is google::protobuf::RepeatedPtrField<SSL_FieldLineSegment>
+   * tbots_robot_msgs is a repeated field in sensor_msg_1
+   */
+
+  // Incorrect
+  vision_msg->set_allocated_ball_state(createBallStateMsg(world.ball()).release());
```
Comment (MathewMacDougall) [inline docs/code-style-guide.md]: Would be good to make this example code more standalone, currently it's just a snippet from real code meaning most people don't know the return type of `createBallStateMsg` for example, or `world.ball()`

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Changes look good, just a nit about the example in the doc

**Review by LiCody — DISMISSED** [review]
lgtm pending comments

File: docs/code-style-guide.md [hunk 4]
```diff
@@ -390,3 +390,22 @@ Some general guidelines when writing tests are:
 ### Protobuf
 
 Protobufs that we define should follow [Google's Protobuf Style Guide](https://developers.google.com/protocol-buffers/docs/style).
+
+* When initializing a protobuf from a unique pointer, avoid passing the `release()`'d pointer to `set_allocated` and `AddAllocated`. Releasing unique pointers makes it too easy to cause memory leaks.
```
Comment (garethellis0) [inline docs/code-style-guide.md]: could we add another point:
```
* The below convention does result in an extra copy operation for the field being added to the proto, but we have decided that this is acceptable to reduce the risk of memory leaks.
```

**Review by garethellis0 — CHANGES_REQUESTED** [review]
As a sanity check, could we run some of these tests under `valgrind`? 

```
bazel run -c dbg --run_under="valgrind --leak-check=full" TARGET_NAME_HERE
```

Just to ensure that we're not accidentally leaking memory anywhere?

**Review by EvanMorcom — DISMISSED** [review]
LGTM, pending Mat's comments 
1 ⭕  ⭕

**Comment by jonathanlew** [comment]
> bazel run -c dbg --run_under="valgrind --leak-check=full"

no memory leaks in any message_translation test, nor sensor_fusion tests

File: docs/code-style-guide.md (hunk 4 above)
Comment (jonathanlew) [reply docs/code-style-guide.md]: added

File: docs/code-style-guide.md (hunk 3 above)
Comment (jonathanlew) [reply docs/code-style-guide.md]: I've added a snippet of the proto definition to make it more clear

**Review by kelvinkoon — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by EvanMorcom — APPROVED** [review]
LGTM 💯

**Review by MathewMacDougall — APPROVED** [review]

**Pull request merged**